src/main/java/org/truffleruby/extra/ConcurrentHashMapNodes.java

```diff
@@ -0,0 +1,300 @@
+import org.truffleruby.builtins.CoreModule;
+import org.truffleruby.builtins.YieldingCoreMethodNode;
+import org.truffleruby.core.basicobject.RubyBasicObject;
+import org.truffleruby.core.hash.RubyHash;
+import org.truffleruby.core.klass.RubyClass;
+import org.truffleruby.core.proc.RubyProc;
+import org.truffleruby.language.Nil;
+import org.truffleruby.language.NotProvided;
+import org.truffleruby.language.Visibility;
+import org.truffleruby.language.objects.AllocationTracing;
+
+import java.util.Map;
+
+@CoreModule(value = "TruffleRuby::ConcurrentHashMap", isClass = true)
+public class ConcurrentHashMapNodes {
+
+    @CoreMethod(names = { "__allocate__", "__layout_allocate__" }, constructor = true, visibility = Visibility.PRIVATE)
+    public abstract static class AllocateNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        protected RubyConcurrentHashMap allocate(RubyClass rubyClass) {
+            final Shape shape = getLanguage().concurrentHashMapShape;
+            final RubyConcurrentHashMap instance = new RubyConcurrentHashMap(rubyClass, shape, null);
+            AllocationTracing.trace(instance, this);
+            return instance;
+        }
+    }
+
+    @CoreMethod(names = "initialize", optional = 1)
+    public abstract static class InitializeNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeNoValue(RubyConcurrentHashMap self, NotProvided options) {
+            return self;
+        }
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeNil(RubyConcurrentHashMap self, Nil options) {
+            return self;
+        }
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeWithOptions(RubyConcurrentHashMap self, RubyHash options) {
+            return self;
+        }
+
+        @Specialization(guards = { "!isNil(options)", "wasProvided(options)" })
+        protected RubyConcurrentHashMap initializeWithOptions(RubyConcurrentHashMap self, RubyBasicObject options) {
+            return self;
+        }
+    }
+
+    @CoreMethod(names = "initialize_copy", required = 1)
+    public abstract static class InitializeCopyNode extends CoreMethodArrayArgumentsNode {
+        @Specialization
+        protected RubyConcurrentHashMap initializeCopy(RubyConcurrentHashMap self, RubyConcurrentHashMap other) {
+            self.concurrentHash.putAll(self.concurrentHash);
+            return self;
+        }
+    }
+
+    @CoreMethod(names = "[]", required = 1)
+    public abstract static class GetIndexNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object getIndex(RubyConcurrentHashMap self, Object key) {
+            final Object value = self.concurrentHash.get(new RubyConcurrentHashMap.Key(key));
+            if (value == null) {
+                return nil;
+            } else {
+                return value;
+            }
```

Comment:
Since we do the several times we should add a helper for converting `nil` back and forth.

Reply:
Right, let's have `nullToNil()` and `nilToNull()` in this class

Reply:
Done in this commit - [`18d59eb` (#2339)](https://github.com/oracle/truffleruby/pull/2339/commits/18d59ebfa0abb17896a47542542124e15c7d1b86)